# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/localstorage/ActorsParent.cpp
# Commit: dc0c88b81dc2
# Full Hash: dc0c88b81dc26894c34ae58fb304dc349dd062f0
# Author: Simon Giesecke <sgiesecke@mozilla.com>
# Date: 2020-12-05 09:38:58
# Regressor Bug: 1678030
# File Overlap Count: 1
# Description:
#   Bug 1678030 - Generalize CreateAndExecuteSingleStepStatement to accept an optional bind functor. r=dom-workers-and-storage-reviewers,ttung
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D97462
# ==============================================================================

diff -r 2fb182664097 -r dc0c88b81dc2 dom/localstorage/ActorsParent.cpp
--- a/dom/localstorage/ActorsParent.cpp	Fri Dec 04 15:10:07 2020 +0000
+++ b/dom/localstorage/ActorsParent.cpp	Fri Dec 04 15:15:31 2020 +0000
@@ -3036,36 +3036,31 @@
                                    ArchivedOriginScope* aArchivedOriginScope) {
   AssertIsOnIOThread();
 
-  // XXX This could use CreateAndExecuteSingleStepStatement from dom/indexedDB
   LS_TRY_INSPECT(
       const auto& stmt,
       ([aArchivedOriginScope,
         &aConnection]() -> Result<nsCOMPtr<mozIStorageStatement>, nsresult> {
         if (aArchivedOriginScope) {
-          LS_TRY_UNWRAP(
-              const auto stmt,
-              MOZ_TO_RESULT_INVOKE_TYPED(
-                  nsCOMPtr<mozIStorageStatement>, aConnection, CreateStatement,
-                  "SELECT "
-                  "total(utf16Length(key) + utf16Length(value)) "
-                  "FROM webappsstore2 "
-                  "WHERE originKey = :originKey "
-                  "AND originAttributes = :originAttributes;"_ns));
-
-          LS_TRY(aArchivedOriginScope->BindToStatement(stmt));
-
-          return stmt;
+          LS_TRY_RETURN(CreateAndExecuteSingleStepStatement<
+                        SingleStepResult::ReturnNullIfNoResult>(
+              aConnection,
+              "SELECT "
+              "total(utf16Length(key) + utf16Length(value)) "
+              "FROM webappsstore2 "
+              "WHERE originKey = :originKey "
+              "AND originAttributes = :originAttributes;"_ns,
+              [aArchivedOriginScope](auto& stmt) -> Result<Ok, nsresult> {
+                LS_TRY(aArchivedOriginScope->BindToStatement(&stmt));
+                return Ok{};
+              }));
         }
 
-        LS_TRY_RETURN(MOZ_TO_RESULT_INVOKE_TYPED(
-            nsCOMPtr<mozIStorageStatement>, aConnection, CreateStatement,
-            "SELECT usage FROM database"_ns));
+        LS_TRY_RETURN(CreateAndExecuteSingleStepStatement<
+                      SingleStepResult::ReturnNullIfNoResult>(
+            aConnection, "SELECT usage FROM database"_ns));
       }()));
 
-  LS_TRY_INSPECT(const bool& hasResult,
-                 MOZ_TO_RESULT_INVOKE(stmt, ExecuteStep));
-
-  LS_TRY(OkIf(hasResult), Err(NS_ERROR_FAILURE));
+  LS_TRY(OkIf(stmt), Err(NS_ERROR_FAILURE));
 
   LS_TRY_RETURN(MOZ_TO_RESULT_INVOKE(stmt, GetInt64, 0));
 }